fil_fungible_token/src/runtime/messaging.rs

```diff
@@ -0,0 +1,160 @@
+use fvm_ipld_encoding::Error as IpldError;
+use fvm_ipld_encoding::RawBytes;
+use fvm_sdk::{actor, send, sys::ErrorNumber};
+use fvm_shared::error::ExitCode;
+use fvm_shared::receipt::Receipt;
+use fvm_shared::METHOD_SEND;
+use fvm_shared::{address::Address, econ::TokenAmount, ActorID};
+use num_traits::Zero;
+use thiserror::Error;
+
+use crate::receiver::types::TokenReceivedParams;
+
+type Result<T> = std::result::Result<T, MessagingError>;
+
+#[derive(Error, Debug)]
+pub enum MessagingError {
+    #[error("fvm syscall error: `{0}`")]
+    Syscall(#[from] ErrorNumber),
+    #[error("address not initialised: `{0}`")]
+    AddressNotInitialised(Address),
+    #[error("ipld serialization error: `{0}`")]
+    Ipld(#[from] IpldError),
+}
+
+/// An abstraction used to send messages to other actors
+pub trait Messaging {
+    /// Call the receiver hook on a given actor, specifying the amount of tokens pending to be sent
+    /// and the sender and receiver
+    ///
+    /// Returns true if the receiver hook is called and exits without error, else returns false
+    fn call_receiver_hook(
+        &self,
+        from: ActorID,
+        to: ActorID,
+        token_value: &TokenAmount,
+        data: &[u8],
+    ) -> Result<Receipt>;
+
+    /// Resolves the given address to it's ID address form
+    fn resolve_id(&self, address: &Address) -> Result<ActorID>;
```

Comment:
I think it's worth documenting the concrete "error" returned when the address is not resolvable, so that caller can catch that one explicitly while passing on others. Ditto below.